lncore/addr.go

```diff
@@ -0,0 +1,319 @@
+		if data.Port == LnDefaultPort {
+			ret[AddrFmtLitIP] = fmt.Sprintf("%s@%s", *data.Pkh, *data.IPAddr)
+		}
+		ret[AddrFmtLitFull] = fmt.Sprintf("%s@%s:%d", *data.Pkh, *data.IPAddr, data.Port)
+	}
+
+	// Hex pubkey stuff
+	if data.Pubkey != nil {
+		ret[AddrFmtPubkey] = fmt.Sprintf("%X", data.Pubkey)
+		if data.IPAddr != nil {
+			if data.Port == LnDefaultPort {
+				ret[AddrFmtPubkeyIP] = fmt.Sprintf("%X@%s", data.Pubkey, *data.IPAddr)
+			}
+			ret[AddrFmtPubkeyIPPort] = fmt.Sprintf("%X@%s:%d", data.Pubkey, *data.IPAddr, data.Port)
+		}
+	}
+
+	// TODO more
+
+	return ret, nil
+
+}
+
+// ParseLnAddrData takes a LnAddr and parses the internal data.  Assumes it's
+// not been force-casted as full checks are done in the ParseLnAddr function.
+func ParseLnAddrData(addr LnAddr) (*LnAddressData, error) {
+
+	addrdata := new(LnAddressData)
+	
+	parts := strings.SplitN(string(addr), "@", 2) // using cast here because reasons
+
+	// First process the identify information.
+	pkdata := strings.SplitN(parts[0], ":", 2)
+	if len(pkdata) == 2 {
+
+		// The pkh is already in the right format for internal storage.
+		addrdata.Pkh = &pkdata[0]		
+		
+		// Now figure out how to parse the full pubkey info.
+		if strings.HasPrefix(pkdata[1], PubkeyBech32Prefix + "1") {
+			_, data, err := bech32.Decode(pkdata[1])
+			if err != nil {
+				return nil, err
+			}
+			addrdata.Pubkey = data
+		} else {
+			data, err := hex.DecodeString(pkdata[1])
+			if err != nil {
+				return nil, err
+			}
+			addrdata.Pubkey = data
+		}
+		
+	} else {
+
+		// If there's only 1 part then there's 3 mutually exclusive options.
+		if strings.HasPrefix(pkdata[0], PkhBech32Prefix + "1") {
+			addrdata.Pkh = &pkdata[0]
+		} else if strings.HasPrefix(pkdata[0], PubkeyBech32Prefix + "1") {
+			_, data, err := bech32.Decode(pkdata[0])
+			if err != nil {
+				return nil, err
+			}
+			addrdata.Pubkey = data
+		} else {
+			data, err := hex.DecodeString(pkdata[0])
+			if err != nil {
+				return nil, err
+			}
+			addrdata.Pubkey = data
+		}
+
+		// Now we figure out what the pkh should be if the full pubkey is specified.
+		if addrdata.Pkh == nil && addrdata.Pubkey != nil {
+			pkh := ConvertPubkeyToBech32Pkh(addrdata.Pubkey)
+			addrdata.Pkh = &pkh
+		}
+		
+	}
+
+	// Now parse the location information if it's there.
+	if len(parts) > 1 {
+		ipdata := strings.SplitN(parts[1], ":", 2)
+
+		// The IP address is already in the right format, probably.
+		addrdata.IPAddr = &ipdata[0]
+
+		// Parse the port if it's there.
+		if len(ipdata) > 1 {
+			port, err := strconv.Atoi(ipdata[1])
+			if err != nil {
+				return nil, err
+			}
+			if port > 65535 {
+				return nil, fmt.Errorf("port number %d not in range", port)
+			}
+			addrdata.Port = uint16(port)
+		} else {
+			addrdata.Port = LnDefaultPort
+		}
+		
+	}
+	
+	return addrdata, nil // TODO
+}
+
+/*
++0      q 	p 	z 	r 	y 	9 	x 	8
++8 	g 	f 	2 	t 	v 	d 	w 	0
++16 	s 	3 	j 	n 	5 	4 	k 	h
++24 	c 	e 	6 	m 	u 	a 	7 	l
+*/
```

Comment:
I guess we could omit this? (Or add this as a comment below the bech32 regex).

Reply:
Yeah that was to help explain the regex.